File: exp8.cpp
#include <iostream>
#include <cmath>
#include <ctime>
#include <vector>
using namespace std;

// 检查是否为素数
bool isPrime(int n) {
    if (n <= 1) return false;
    for (int i = 2; i <= sqrt(n); i++)
        if (n % i == 0) return false;
    return true;
}

// 计算最大公约数
int gcd(int a, int b) {
    return b == 0 ? a : gcd(b, a % b);
}

// 生成密钥对
void generateKeys(int p, int q, int &n, int &e, int &d) {
    n = p * q;
    int phi = (p-1)*(q-1);
    
    // 选择公钥e (1 < e < phi，且与phi互质)
    for (e = 2; e < phi; e++)
        if (gcd(e, phi) == 1) break;
    
    // 计算私钥d (d*e ≡ 1 mod phi)
    for (d = 2; d < phi; d++)
        if ((d * e) % phi == 1) break;
}

// 加密函数
int encrypt(int msg, int e, int n) {
    int result = 1;
    for (int i = 0; i < e; i++)
        result = (result * msg) % n;
    return result;
}

// 解密函数
int decrypt(int cipher, int d, int n) {
    int result = 1;
    for (int i = 0; i < d; i++)
        result = (result * cipher) % n;
    return result;
}

// 因式分解破解
void factorize(int n, int &p, int &q) {
    clock_t start = clock();
    for (p = 2; p <= sqrt(n); p++) {
        if (n % p == 0 && isPrime(p)) {
            q = n / p;
            if (isPrime(q)) break;
        }
    }
    clock_t end = clock();
    double time = double(end - start) / CLOCKS_PER_SEC;
    cout << "破解耗时: " << time << "秒" << endl;
}

int main() {
    cout << "======== RSA加解密实验 ========" << endl;
    
    // 实验1：完整的RSA流程演示（32位）
    {
        cout << "\n[实验1] 32位RSA完整流程演示" << endl;
        int p = 61, q = 53; // 两个小素数
        int n, e, d;
        
        generateKeys(p, q, n, e, d);
        cout << "公钥(e,n): (" << e << ", " << n << ")" << endl;
        cout << "私钥(d,n): (" << d << ", " << n << ")" << endl;
        
        int msg = 123; // 原始消息
        cout << "原始消息: " << msg << endl;
        
        int cipher = encrypt(msg, e, n);
        cout << "加密结果: " << cipher << endl;
        
        int decrypted = decrypt(cipher, d, n);
        cout << "解密结果: " << decrypted << endl;
    }
    
    // 实验2：不同密钥长度的破解时间比较
    cout << "\n[实验2] 不同密钥长度的破解时间比较" << endl;
    vector<pair<int, string>> testCases = {
        {32, "32位 (4-5位十进制)"},
        {64, "64位 (19-20位十进制)"},
        {128, "128位 (38-39位十进制)"},
        {256, "256位 (77-78位十进制)"},
        {512, "512位 (154-155位十进制)"}
    };
    
    for (auto &test : testCases) {
        cout << "\n测试密钥长度: " << test.second << endl;
        
        // 生成两个素数
        int bits = test.first / 2;
        int min = pow(2, bits-1);
        int max = pow(2, bits) - 1;
        
        int p = 0, q = 0;
        while (!isPrime(p)) p = rand() % (max - min + 1) + min;
        while (!isPrime(q)) q = rand() % (max - min + 1) + min;
        
        int n = p * q;
        cout << "模数n: " << n << endl;
        
        int found_p, found_q;
        factorize(n, found_p, found_q);
        
        if (found_p * found_q == n) {
            cout << "破解成功! 因数: " << found_p << " 和 " << found_q << endl;
        } else {
            cout << "破解失败!" << endl;
        }
    }
    system("pause");
    return 0;
}

File: exp8.py
import math
import random
import time
from multiprocessing import Process, Queue


def is_prime(n):
    """检查是否为素数"""
    if n <= 1:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

def generate_prime(min_val, max_val):
    """生成指定范围内的素数"""
    while True:
        p = random.randint(min_val, max_val)
        if is_prime(p):
            return p

def gcd(a, b):
    """计算最大公约数"""
    while b != 0:
        a, b = b, a % b
    return a

def modinv(e, phi):
    """计算模逆元"""
    for d in range(2, phi):
        if (d * e) % phi == 1:
            return d
    return None

def generate_keys(p, q):
    """生成RSA密钥对"""
    n = p * q
    phi = (p-1) * (q-1)
    
    # 选择公钥e
    for e in range(2, phi):
        if gcd(e, phi) == 1:
            break
    
    # 计算私钥d
    d = modinv(e, phi)
    
    return (e, n), (d, n)

def encrypt(msg, public_key):
    """加密函数"""
    e, n = public_key
    return pow(msg, e, n)

def decrypt(cipher, private_key):
    """解密函数"""
    d, n = private_key
    return pow(cipher, d, n)

def factorize(n, result_queue):
    """因式分解函数"""
    start_time = time.time()
    
    for p in range(2, int(math.sqrt(n)) + 1):
        if n % p == 0 and is_prime(p):
            q = n // p
            if is_prime(q):
                result_queue.put((p, q, time.time() - start_time))
                return
    
    result_queue.put((None, None, time.time() - start_time))

def crack_rsa(n, timeout=120):
    """破解RSA"""
    print(f"开始破解n={n}...")
    result_queue = Queue()
    p = Process(target=factorize, args=(n, result_queue))
    p.start()
    
    # 等待结果或超时
    p.join(timeout)
    
    if p.is_alive():
        p.terminate()
        p.join()
        return None, None, None, "失败（超过120秒）"
    
    p, q, time_used = result_queue.get()
    if p and q:
        return p, q, time_used, "成功"
    else:
        return None, None, time_used, "失败（未找到因数）"

def main():
    print("======== RSA加解密实验========")
    
    # 实验1：完整的RSA流程演示（8位）
    print("\n[实验1] 8位RSA完整流程")
    p = 23
    q = 29
    public_key, private_key = generate_keys(p, q)
    print(f"公钥(e,n): {public_key}")
    print(f"私钥(d,n): {private_key}")
    
    msg = 42  # 原始消息必须小于n
    print(f"原始消息: {msg}")
    
    cipher = encrypt(msg, public_key)
    print(f"加密结果: {cipher}")
    
    decrypted = decrypt(cipher, private_key)
    print(f"解密结果: {decrypted}")
    
    # 实验2：不同密钥长度的破解时间比较
    print("\n[实验2] 不同密钥长度的破解时间比较（最多等待120秒）")
    test_cases = [
        (16, "16位"),
        (20, "20位"),
        (24, "24位"),
        (26, "26位"),
        (28, "28位"),
        (30, "30位"),
        (32, "32位")
    ]
    
    for bits, description in test_cases:
        print(f"\n测试密钥长度: {description}")
        
        # 生成两个素数
        min_val = 10 ** (bits//4)  # 调整范围使n不会太大
        max_val = 10 ** (bits//4 + 1)
        
        p = generate_prime(min_val, max_val)
        q = generate_prime(min_val, max_val)
        n = p * q
        print(f"模数n: {n} (位数: {len(str(n))})")
        
        # 尝试破解
        found_p, found_q, time_used, status = crack_rsa(n)
        
        if status.startswith("成功"):
            print(f"破解{status}! 因数: {found_p} 和 {found_q}")
        else:
            print(f"破解{status}")
        
        print(f"耗时: {time_used:.2f}秒" if time_used is not None else "超过120秒")

if __name__ == "__main__":
    main()